illumos-utils/src/opte/illumos/port_manager.rs

```diff
@@ -454,31 +351,239 @@ impl PortManager {
+        let vpc_cfg = VpcCfg {
+            ip_cfg: IpCfg::Ipv4(Ipv4Cfg {
+                vpc_subnet: opte_subnet,
+                private_ip,
+                gateway_ip,
+                snat,
+                external_ips: external_ip,
+            }),
+            guest_mac: MacAddr::from(mac.into_array()),
+            gateway_mac: MacAddr::from(gateway.mac.into_array()),
+            vni,
+            phys_ip: self.inner.underlay_ip.into(),
+            boundary_services,
+            // TODO-remove: Part of the external IP hack.
+            //
+            // NOTE: This value of this flag is irrelevant, since the driver
+            // always overwrites it. The field itself is used in the `oxide-vpc`
+            // code though, to determine how to set up the ARP layer, which is
+            // why it's still here.
+            proxy_arp_enable: true,
+            phys_gw_mac: Some(MacAddr::from(
+                self.inner.gateway_mac.into_array(),
+            )),
+            // TODO-completeness (#2153): Plumb domain search list
+            domain_list: vec![],
+        };
+
+        self.create_port_inner(
+            instance_id,
+            PortType::Guest { slot: nic.slot },
+            None,
+            vpc_cfg,
+            firewall_rules,
+            external_ip.is_some(),
+        )
+    }
+
+    pub fn create_svc_port(
+        &self,
+        svc_id: Uuid,
+        svc_name: &str,
+        mac: MacAddr6,
+        external_ip: IpAddr,
+        vni: u32,
+        firewall_rules: &[VpcFirewallRule],
+    ) -> Result<(Port, PortTicket), Error> {
+        // TODO: Assume each service zone has a single port.
+        let port_name = format!("{}_{}0", XDE_LINK_PREFIX, svc_name);
+
+        // TODO: Arbitrary subnet, should come from a "service" VPC.
+        let vpc_subnet: Ipv4Network = "10.0.0.0/24".parse().unwrap();
+        // Use first non-reserved IP, see RFD 21, section 2.2, table 1.
+        let private_ip = vpc_subnet.iter().nth(5).unwrap();
+        let gateway = Gateway::from_subnet(&vpc_subnet.into());
+
+        // Convert to OPTE types
+        let vpc_subnet = Ipv4Cidr::from(vpc_subnet);
+        let private_ip = private_ip.into();
+        let gateway_ip = match gateway.ip {
+            IpAddr::V4(ip) => ip.into(),
+            IpAddr::V6(_) => unreachable!("expected IPv4 gateway"),
+        };
+        let gateway_mac = MacAddr::from(gateway.mac.into_array());
+        let guest_mac = MacAddr::from(mac.into_array());
+        let external_ip = match external_ip {
+            IpAddr::V4(ip) => ip.into(),
+            IpAddr::V6(_) => {
+                return Err(opte_ioctl::Error::InvalidArgument(String::from(
+                    "IPv6 is not yet supported for external addresses",
+                ))
+                .into());
+            }
+        };
+
+        let snat = SNat4Cfg {
+            external_ip,
+            // TODO: Assumes IP is exlusive to this service.
```

Comment:
typo here, plus: isn't this true (it is at least for IPv6)?

Reply:
It looks like this uses the same external IP to create the `SNat4Cfg` and also passes that as the `external_ips` field of the `Ipv4Cfg`. That's "fine" in the sense that OPTE will always use the latter if it exists, ignoring the former. I don't see a difference between IPv4 and IPv6 here.

One question I have is where these addresses come from. Eventually, operators will provide the address lots they want to assign to the rack. We'll then take one of these for Nexus, possibly a specific address. It's not part of this PR, but we will want to store that in the DB and explicitly assign the address to Nexus, along with the entire port range. That would realistically reserve the whole port range, since this must allow incoming connections.

I actually think we should consider _not_ providing an `SNatCfg` at all here. It plays no role, since OPTE will ignore it, and I think it might confuse things more.

Reply:
Removed the snat config for service ports which exclusively use external IPs right now.